staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -684,20 +685,25 @@ func (r *Request) newStreamWatcher(resp *http.Response) (watch.Interface, error)
 
 // updateURLMetrics is a convenience function for pushing metrics.
 // It also handles corner cases for incomplete/invalid request data.
-func updateURLMetrics(ctx context.Context, req *Request, resp *http.Response, err error) {
+func updateURLMetrics(ctx context.Context, finalAttempt bool, req *Request, resp *http.Response, err error) {
```

Comment:
I feel is cleaner to not overload this function, this is also changes the metrics to get only the final request result, it may be helpful to record all the status a request goes through i.e. 
5 retries with 4x500 status and one last 200
Can we move the retry metric just inside the retry logic?

Reply:
> I feel is cleaner to not overload this function,
I refactored a bit, want to make sure the code and url sanitization logic is shared. 

> it may be helpful to record all the status a request goes through i.e.
> 5 retries with 4x500 status and one last 200

yes, so with this PR, we will see the following:
- `rest_client_request_retry_total`  will be at  {4, code=500} 
- `rest_client_requests_total`  will be at  {1, code=200}

are you suggesting the following?
- `rest_client_request_retry_total`  will be at  {4, code=500} 
- `rest_client_requests_total`  will be at  {4, code=500}  {1, code=200}

> Can we move the retry metric just inside the retry logic?
I think it's cleaner to keep the metrics call co-located, also the retry logic in `Before` or `After` are not aware if this is the final retry.

Reply:
> rest_client_request_retry_total will be at {4, code=500}
> rest_client_requests_total will be at {4, code=500} {1, code=200}

reflects the reality from the apiserver, there are 4 client requests at the end of the cycle, 3 of them were retries, although , from the client only one request was created ... I think that it depends who is the observer here, what are the best practices here  @dgrisonnet ?

Reply:
I think we should still have `rest_client_requests_total` account for all the requests made by the client. With that metric, you want to know how many requests were made by the client library to the server and even if a request might end up succeeding after being retried, the original failures will impact both the apiserver and the client.

One thing we have to keep in mind is that with these metrics we want to monitor the connection between the client and the server, not the actual result of the use of the client library. So whether the library call made by its consumer end up succeeding or not, shouldn't be something we care about from a `rest_client` metrics perspective.

Let me know if that make sense to you, but IMHO I don't think monitoring retries is useful here.

Reply:
> I think we should still have rest_client_requests_total account for all the requests made by the client. With that metric, you want to know how many requests were made by the client library to the server and even if a request might end up succeeding after being retried, the original failures will impact both the apiserver and the client.

@dgrisonnet  if we think about the user experience, the user is taking action based on the result of the last attempt(s), not the previous retries.  `rest_client_requests_total` counts the result(s) that the user is interested in, and it can be used as an indicator for QoS, a request with 4 `{5xx, Retry-After: 1}` and a final `200` is still a success to the user.

`rest_client_request_retry_total` can be used to understand the degree of retries - are we seeing excessive number of retries, and such.
if we sum up the two metrics, we get the total number of attempts sent to the server.

if `rest_client_requests_total` counts the results from the non-final attempts as well, we can't say whether an error code `5xx` is because the server asked us to retry, or an actual failure case. thoughts?